File: script/make/protocol/js/BubbleProtocol.js
export default class BubbleProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 12101: {

                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 12101: {
                // 气泡列表
                const data = [];
                let dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--dataLength >= 0) {
                    // 
                    // 气泡ID
                    const dataDataBubbleId = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 过期时间
                    const dataDataExpireTime = view.getUint32(offset, false);
                    offset = offset + 4;
                    // object
                    const dataData = {"bubbleId": dataDataBubbleId, "expireTime": dataDataExpireTime};
                    // add
                    data.push(dataData);
                }
                return data;
            }
            case 12102: {
                // 气泡ID列表
                const data = [];
                let dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--dataLength >= 0) {
                    // 气泡ID
                    const dataData = view.getUint32(offset, false);
                    offset = offset + 4;
                    // add
                    data.push(dataData);
                }
                return data;
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}

File: script/make/protocol/lua/BubbleProtocol.lua
BubbleProtocol = {}

function BubbleProtocol.encode(offset, protocol, data)
    if protocol == 12101 then
        local table = {}

        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function BubbleProtocol.decode(offset, protocol, bytes)
    if protocol == 12101 then
        -- 气泡列表
        local data = {}
        local dataLength = string.unpack(">I2", bytes, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- 
            -- 气泡ID
            local dataDataBubbleId = string.unpack(">I4", bytes, offset)
            offset = offset + 4
            -- 过期时间
            local dataDataExpireTime = string.unpack(">I4", bytes, offset)
            offset = offset + 4
            -- object
            local dataData = {bubbleId = dataDataBubbleId, expireTime = dataDataExpireTime}
            data[dataIndex] = dataData
        end
        return data
    elseif protocol == 12102 then
        -- 气泡ID列表
        local data = {}
        local dataLength = string.unpack(">I2", bytes, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- 气泡ID
            local dataData = string.unpack(">I4", bytes, offset)
            offset = offset + 4
            data[dataIndex] = dataData
        end
        return data
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end